src/backends/r1cs/mod.rs

```
@@ -0,0 +1,489 @@
+            for lc in constraint.as_array() {
+                for var in lc.terms.keys() {
+                    written_vars.insert(var.index);
+                }
+            }
+        }
+
+        // check if every cell vars end up being a cell var in the circuit or public output
+        for index in 0..self.witness_vars.len() {
+            if !written_vars.contains(&index) {
+                // check if outputs contains the cell var that has the same index
+                if !self.public_outputs.iter().any(|&x| x.index == index) {
+                    panic!("there's a bug in the circuit_writer, some cellvar does not end up being a cellvar in the circuit!");
+                }
+            }
+        }
+
+        self.finalized = true;
+
+        Ok(())
+    }
+
+    /// Generate the witnesses
+    /// This process should check if the constraints are satisfied.
+    fn generate_witness(
+        &self,
+        witness_env: &mut crate::witness::WitnessEnv<Self::Field>,
+    ) -> crate::error::Result<Self::GeneratedWitness> {
+        if !self.finalized {
+            panic!("the circuit is not finalized yet!");
+        }
+
+        // generate witness through witness vars vector
+        let mut witness = self.witness_vars.iter().enumerate().map(|(index, val)| {
+            self.compute_val(witness_env, val, index)
+        }).collect::<crate::error::Result<Vec<Fr>>>()?;
+
+        // The original vars of public outputs are not part of the constraints
+        // so we need to compute them separately
+        for var in &self.public_outputs {
+            let val = self.compute_var(witness_env, *var)?;
+            witness[var.index] = val;
+        }
+
+        for constraint in &self.constraints {
+            // assert a * b = c
+            assert_eq!(
+                constraint.a.evaluate(&witness) * constraint.b.evaluate(&witness), 
+                constraint.c.evaluate(&witness)
+            );
+        }
```

we should return a nice error to the user here, like we do in kimchi:

```rust
return Err(Error::new(
                                "runtime",
                                ErrorKind::InvalidWitness(row),
                                debug_info.span,
                            ));
```